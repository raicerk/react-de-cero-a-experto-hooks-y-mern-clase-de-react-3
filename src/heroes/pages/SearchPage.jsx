import { useLocation, useNavigate } from "react-router-dom";
import { useForm } from "../../hooks/useform";
import { HeroCard } from "../components/HeroCard";
import queryString from "query-string";
import { getHeroesByName } from "../helpers/getHeroesByName";

export const SearchPage = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const { q = "" } = queryString.parse(location.search);

  const heroes = getHeroesByName(q);

  const { searchText, onInputChange } = useForm({ searchText: q });

  const onSearchSubmit = (event) => {
    event.preventDefault();
    navigate(`?q=${searchText}`);
  };

  return (
    <>
      <h1>Search</h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <h4>Searching</h4>
          <hr />
          <form onSubmit={onSearchSubmit}>
            <input
              type="text"
              placeholder="Search a hero"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={searchText}
              onChange={onInputChange}
            />
            <button className="btn btn-outline-primary" type="submit">
              Search
            </button>
          </form>
        </div>
        <div className="col-7">
          <h4>Results</h4>
          <hr />

          {q === "" ? (
            <div className="alert alert-primary">Search a hero:</div>
          ) : (
            heroes.length === 0 && (
              <div className="alert alert-danger">
                No hero with <b>{q}</b>
              </div>
            )
          )}

          {heroes.map((hero) => {
            return <HeroCard key={hero.id} {...hero} />;
          })}
        </div>
      </div>
    </>
  );
};
