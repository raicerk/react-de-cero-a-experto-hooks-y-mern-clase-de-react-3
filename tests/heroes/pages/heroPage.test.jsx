import { fireEvent, render, screen } from "@testing-library/react";
import { HeroPage } from "../../../src/heroes/pages/HeroPage";
import rrd, { MemoryRouter } from "react-router-dom";

const mockNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({
    heroId: "dc-batman",
  }),
  useNavigate: () => mockNavigate,
}));

describe("Prueba del componente <heroPage/>", () => {
  test("Mostrar componente por defecto", () => {
    const { container } = render(
      <MemoryRouter>
        <HeroPage />
      </MemoryRouter>
    );

    expect(screen.getByRole("heading", { level: 3 }).innerHTML).toContain(
      "Batman"
    );
    expect(container).toMatchSnapshot();
  });

  test("Debe llamarse a la funcion de volver atrÃ¡s", () => {
    render(
      <MemoryRouter>
        <HeroPage />
      </MemoryRouter>
    );

    const button = screen.getByRole("button");
    fireEvent.click(button);
    expect(mockNavigate).toBeCalled();
  });

  test("No debe cargar el componente por no recibir el useParam correcto", () => {
    jest.spyOn(rrd, "useParams").mockImplementation(() => {
      return {
        heroId: "",
      };
    });

    const { container } = render(
      <MemoryRouter>
        <HeroPage />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });
});
