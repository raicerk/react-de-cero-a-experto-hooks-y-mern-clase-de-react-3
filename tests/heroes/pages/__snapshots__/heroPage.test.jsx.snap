// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba del componente <heroPage/> Mostrar componente por defecto 1`] = `
<div>
  <div
    class="row mt-5"
  >
    <div
      class="col-4"
    >
      <img
        alt="Batman"
        class="img-thumbnail"
        src="/assets/dc-batman.jpg"
      />
    </div>
    <div
      class="col-8"
    >
      <h3>
        Batman
      </h3>
      <ul
        class="list-group list-group-flush"
      >
        <li
          class="list-group-item"
        >
          <b>
            Alter ego:
          </b>
           
          Bruce Wayne
        </li>
        <li
          class="list-group-item"
        >
          <b>
            Publisher:
          </b>
           
          DC Comics
        </li>
        <li
          class="list-group-item"
        >
          <b>
            First appearance:
          </b>
           
          Detective Comics #27
        </li>
      </ul>
      <h5
        class="mt-3"
      >
         Characters
      </h5>
      <p>
        Bruce Wayne
      </p>
      <button
        class="btn btn-outline-primary"
      >
        Regresar
      </button>
    </div>
  </div>
</div>
`;

exports[`Prueba del componente <heroPage/> No debe cargar el componente por no recibir el useParam correcto 1`] = `<div />`;
